generated/serialization.h

```diff
@@ -0,0 +1,80 @@
+#ifndef DTOGEN_SERIALIZATION_H
+#define DTOGEN_SERIALIZATION_H
+
+#include <memory>
+#include <optional>
+
+#include "nlohmann/json.hpp"
+
+// TODO: move elsewhere
+#ifdef __clang__
+#define UNREACHABLE_CODE __builtin_unreachable();
+#elif __GNUC__
+#define UNREACHABLE_CODE __builtin_unreachable();
+#elif _MSC_VER
+#define UNREACHABLE_CODE __assume(0);
+#endif
+
+// TODO: move elsewhere
+#include <stdexcept>
+namespace org {
+namespace openapitools {
+namespace model {
+class invalid_enum_value : public std::runtime_error {
+    public:
+        using std::runtime_error::runtime_error;
+};
+}  // namespace org
+}  // namespace openapitools
+}  // namespace model
+
+
+namespace nlohmann {
+
+template <typename T>
+struct adl_serializer<std::optional<T>> {
+    static void to_json(json& j, const std::optional<T>& opt) {
+        opt == std::nullopt ? j = nullptr : j = *opt;
+    }
+
+    static void from_json(const json& j, std::optional<T>& opt) {
+        j.is_null() ? opt = std::nullopt : opt = j.get<T>();
+    }
+};
+
+template <typename T>
+struct adl_serializer<std::shared_ptr<T>> {
+    static void to_json(json& j, const std::shared_ptr<T>& ptr) {
+        j = ptr ? *ptr : nullptr;
+    }
+
+    static void from_json(const json& j, std::shared_ptr<T>& ptr) {
+        if (j.is_null()) {
+            ptr.reset();
+        } else {
+            ptr = std::make_shared<T>(j.get<T>());
+        }
+    }
+};
+
+}  // namespace nlohmann
+
+namespace org {
+namespace openapitools {
+namespace model {
+namespace detail {
+
+template <typename T>
+void to_optional(const nlohmann::json& j, const char* key, T& target) {
+    if (j.contains(key)) {
+        j.at(key).get_to(target);
+    }
```

Comment:
Probably its a good idea to explicitly set `target` to `std::nullopt`. Just in case somebody tried to out-smart  the parser code and set certain optional members before parsing. I.e.

```C++
if (j.contains(key)) {
  // ...
} else {
  target = std::nullopt;
}
```